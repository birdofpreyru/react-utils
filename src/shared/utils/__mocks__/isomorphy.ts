/**
 * Mockup of the isomorphy module.
 */
/* global jest */

import * as IsomorphyM from '../isomorphy';

const MOCK_BUILD_TIMESTAMP = 'Wed, 29 Nov 2017 07:40:00 GMT';

const mock = jest.requireActual<typeof IsomorphyM>('../isomorphy');

mock.buildTimestamp = () => MOCK_BUILD_TIMESTAMP;

module.exports = mock;
