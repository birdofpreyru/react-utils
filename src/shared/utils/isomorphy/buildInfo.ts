// Encapsulates access to "Build Info" data.

// BEWARE: This should match the type of build info object generated by
// Webpack build (see "/config/webpack/app-base.js"), and currently this
// match is not checked automatically.
export type BuildInfoT = {
  key: string;
  publicPath: string;
  timestamp: string;
  useServiceWorker: boolean;
};

// Depending on the build mode & environment, BUILD_INFO is either a global
// variable defined at the app launch, or it is replaced by the actual value
// by the Webpack build.
declare const BUILD_INFO: BuildInfoT | undefined;

let buildInfo: BuildInfoT | undefined;

// On the client side "BUILD_INFO" should be injected by Webpack. Note, however,
// that in test environment we may need situations were environment is mocked as
// client-side, although no proper Webpack compilation is executed, thus no info
// injected; because of this we don't do a hard environment check here.
if (typeof BUILD_INFO !== 'undefined') buildInfo = BUILD_INFO;

/**
 * In scenarious where "BUILD_INFO" is not injected by Webpack (server-side,
 * tests, etc.) we expect the host codebase to explicitly set it before it is
 * ever requested. As a precaution, this function throws if build info has been
 * set already, unless `force` flag is explicitly set.
 * @param info
 * @param force
 */
export function setBuildInfo(info?: BuildInfoT, force = false): void {
  if (buildInfo !== undefined && !force) {
    throw Error('"Build Info" is already initialized');
  }
  buildInfo = info;
}

/**
 * Returns "Build Info" object; throws if it has not been initialized yet.
 * @returns
 */
export function getBuildInfo(): BuildInfoT {
  if (buildInfo === undefined) {
    throw Error('"Build Info" has not been initialized yet');
  }
  return buildInfo;
}
