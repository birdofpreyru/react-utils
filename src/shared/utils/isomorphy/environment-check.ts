// Checks for client- vs. server-side environment detection.

/**
 * `true` within client-side environment (browser), `false` at server-side.
 */
export const IS_CLIENT_SIDE: boolean = typeof process !== 'object'
  || !process.versions?.node
  || !!global.REACT_UTILS_FORCE_CLIENT_SIDE;

/**
 * `true` within the server-side environment (node), `false` at client-side.
 */
export const IS_SERVER_SIDE: boolean = !IS_CLIENT_SIDE;
