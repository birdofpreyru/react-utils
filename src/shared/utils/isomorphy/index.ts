import { getBuildInfo } from './buildInfo';
import { IS_CLIENT_SIDE, IS_SERVER_SIDE } from './environment-check';

/**
 * @ignore
 * @return {string} Code mode: "development" or "production".
 */
function getMode() {
  return process.env.NODE_ENV;
}

/**
 * Returns `true` if development version of the code is running;
 * `false` otherwise.
 */
export function isDevBuild(): boolean {
  return getMode() === 'development';
}

/**
 * Returns `true` if production build of the code is running;
 * `false` otherwise.
 */
export function isProdBuild(): boolean {
  return getMode() === 'production';
}

/**
 * Returns build timestamp of the front-end JS bundle.
 * @return ISO date/time string.
 */
export function buildTimestamp(): string {
  return getBuildInfo().timestamp;
}

export { IS_CLIENT_SIDE, IS_SERVER_SIDE, getBuildInfo };
