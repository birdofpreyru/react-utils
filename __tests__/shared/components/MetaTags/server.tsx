/**
 * Tests for the server-side MetaTags integration.
 */

import { noop } from 'lodash';
import type { ComponentType } from 'react';
import supertest from 'supertest';

import serverFactory from 'server/server';
import { setBuildInfo } from 'utils/isomorphy/buildInfo';

import Application, { MODES } from './__assets__/testcase/app';
import Application2, { MODES as MODES2 } from './__assets__/testcase/app2';

const WEBPACK_CONFIG = {
  context: `${__dirname}/__assets__/testcase`,
  output: {
    path: '/path/to/assets',
    publicPath: '/test/public/path',
  },
};

async function baseTest(Component: ComponentType) {
  const server = supertest(
    await serverFactory(WEBPACK_CONFIG, {
      Application: Component,
      beforeRender: () => ({
        initialState: { domain: 'https://sample.domain' },
      }),
      logger: {
        debug: noop,
        error: noop,
        info: noop,
        log: noop,
        warn: noop,
      },
      maxSsrRounds: 3,
      ssrTimeout: Number.MAX_VALUE,
    }),
  );
  await server.get('/').expect(200)
    .expect((res) => {
      expect(res.text).toMatchSnapshot();
    });
}

afterEach(() => {
  setBuildInfo(undefined, true);
});

test('Basic tags, no override', async () => baseTest(
  () => <Application mode={MODES.BASIC_NO_OVERRIDE} />,
));

test('Basic tags, with override', async () => baseTest(
  () => <Application mode={MODES.BASIC_WITH_OVERRIDE} />,
));

test('All tags, with override', async () => baseTest(
  () => <Application mode={MODES.ALL_TAGS_WITH_OVERRIDE} />,
));

describe('Using <MetaTags> children feature', () => {
  test('Basic tags, no override', async () => baseTest(
    () => <Application2 mode={MODES2.BASIC_NO_OVERRIDE} />,
  ));

  test('Basic tags, with override', async () => baseTest(
    () => <Application2 mode={MODES2.BASIC_WITH_OVERRIDE} />,
  ));
});
