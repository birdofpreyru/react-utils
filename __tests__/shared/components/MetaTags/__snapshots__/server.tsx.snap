// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`All tags, with override 1`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true">Title from Component</title>
            <meta data-rh="true" content="Component description" name="description"/><meta data-rh="true" content="summary_large_image" name="twitter:card"/><meta data-rh="true" content="SOCIAL_TITLE" name="twitter:title"/><meta data-rh="true" content="SOCIAL_DESCRIPTION" name="twitter:description"/><meta data-rh="true" content="https://sample.domain/THUMBNAIL_PATH" name="twitter:image"/><meta data-rh="true" content="@SITE_NAME" name="twitter:site"/><meta data-rh="true" content="SOCIAL_TITLE" name="og:title"/><meta data-rh="true" content="https://sample.domain/THUMBNAIL_PATH" name="og:image"/><meta data-rh="true" content="SOCIAL_TITLE" name="og:image:alt"/><meta data-rh="true" content="SOCIAL_DESCRIPTION" name="og:description"/><meta data-rh="true" content="SITE_NAME" name="og:sitename"/><meta data-rh="true" content="PAGE_URL" name="og:url"/>
            <meta name="theme-color" content="#FFFFFF">
            
            
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghgc5WRH69eBKHiBZQmMDPrDIZjS0mo8ysGCjayxGzD6WmL8eLQdJGGbkbV9YPqFyAc1uVFm6WcpMG4mD/Npa7cqgK6UVPOjCVTlFbZPBYhN1gUlx9R+UmuUKa6MNZMcJG">
          </head>
          <body>
            
            <div id="react-view"><div><p>Hello World!</p><div><p>A dummy internal component.</p></div></div></div>
            
            
          </body>
        </html>"
`;

exports[`Basic tags, no override 1`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true">Application Title</title>
            <meta data-rh="true" content="Application Description" name="description"/><meta data-rh="true" content="summary_large_image" name="twitter:card"/><meta data-rh="true" content="Application Title" name="twitter:title"/><meta data-rh="true" content="Application Description" name="twitter:description"/><meta data-rh="true" content="Application Title" name="og:title"/><meta data-rh="true" content="Application Description" name="og:description"/>
            <meta name="theme-color" content="#FFFFFF">
            
            
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghgc5WRH69eBKHiBZQmMDPrDIZjS0mo8ysGCjayxGzD6WmL8eLQdJGGbkbV9YPqFyAc1uVFm6WcpMG4mD/Npa7cqgK6UVPOjCVTlFbZPBYhN1gUlx9R+UmuUKa6MNZMcJG">
          </head>
          <body>
            
            <div id="react-view"><div><p>Hello World!</p></div></div>
            
            
          </body>
        </html>"
`;

exports[`Basic tags, with override 1`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true">Title from Component</title>
            <meta data-rh="true" content="Component description" name="description"/><meta data-rh="true" content="summary_large_image" name="twitter:card"/><meta data-rh="true" content="Title from Component" name="twitter:title"/><meta data-rh="true" content="Component description" name="twitter:description"/><meta data-rh="true" content="Title from Component" name="og:title"/><meta data-rh="true" content="Component description" name="og:description"/>
            <meta name="theme-color" content="#FFFFFF">
            
            
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghgc5WRH69eBKHiBZQmMDPrDIZjS0mo8ysGCjayxGzD6WmL8eLQdJGGbkbV9YPqFyAc1uVFm6WcpMG4mD/Npa7cqgK6UVPOjCVTlFbZPBYhN1gUlx9R+UmuUKa6MNZMcJG">
          </head>
          <body>
            
            <div id="react-view"><div><p>Hello World!</p><div><p>A dummy internal component.</p></div></div></div>
            
            
          </body>
        </html>"
`;

exports[`Using <MetaTags> children feature Basic tags, no override 1`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true">Application Title</title>
            <meta data-rh="true" content="Application Description" name="description"/><meta data-rh="true" content="summary_large_image" name="twitter:card"/><meta data-rh="true" content="Application Title" name="twitter:title"/><meta data-rh="true" content="Application Description" name="twitter:description"/><meta data-rh="true" content="Application Title" name="og:title"/><meta data-rh="true" content="Application Description" name="og:description"/>
            <meta name="theme-color" content="#FFFFFF">
            
            
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghgc5WRH69eBKHiBZQmMDPrDIZjS0mo8ysGCjayxGzD6WmL8eLQdJGGbkbV9YPqFyAc1uVFm6WcpMG4mD/Npa7cqgK6UVPOjCVTlFbZPBYhN1gUlx9R+UmuUKa6MNZMcJG">
          </head>
          <body>
            
            <div id="react-view"><div><p>Hello World!</p></div></div>
            
            
          </body>
        </html>"
`;

exports[`Using <MetaTags> children feature Basic tags, with override 1`] = `
"<!DOCTYPE html>
        <html lang="en">
          <head>
            
            <title data-rh="true">Title from Component - Application Title</title>
            <meta data-rh="true" content="Component description - Application Description" name="description"/><meta data-rh="true" content="summary_large_image" name="twitter:card"/><meta data-rh="true" content="Title from Component - Application Title" name="twitter:title"/><meta data-rh="true" content="Component description - Application Description" name="twitter:description"/><meta data-rh="true" content="Title from Component - Application Title" name="og:title"/><meta data-rh="true" content="Component description - Application Description" name="og:description"/>
            <meta name="theme-color" content="#FFFFFF">
            
            
            
            <meta charset="utf-8">
            <meta
              content="width=device-width,initial-scale=1.0"
              name="viewport"
            >
            <meta itemprop="drpruinj" content="MDEyMzQ1Njc4OTAxMjM0NTY3ODkwMTIzNDU2Nzg5MDEMKtsRsLh8R6HHpitZ/eghgc5WRH69eBKHiBZQmMDPrDIZjS0mo8ysGCjayxGzD6WmL8eLQdJGGbkbV9YPqFyAc1uVFm6WcpMG4mD/Npa7cqgK6UVPOjCVTlFbZPBYhN1gUlx9R+UmuUKa6MNZMcJG">
          </head>
          <body>
            
            <div id="react-view"><div><p>Hello World!</p><div><p>A dummy internal component.</p></div></div></div>
            
            
          </body>
        </html>"
`;
